src/cascade.js

```diff
@@ -0,0 +1,30 @@
+import _ from 'lodash/fp'
+import { getTypePropOrError } from './types'
+
```

Comment:
This file would benefit from an explanation of what is cascading and why we need it. It would also benefit from providing the type signature of every one of the parameters, as well as the expected output.

Reply:
Also, how is this different from futil’s cascade? perhaps we could use another name. Or maybe it makes sense to use the same name, but then explaining why seems like a good idea to me.

Reply:
Docs explaining usage aren't here yet because the unit tests aren't done yet (so the top level API might still change). For example, this might end up exposed as methods on the tree itself as `cascadeFrom` and `cascadeTo`. But definitely noted 👍 

For naming, this is unrelated to futil cascade. It's called `cascade` as a carry-over from the DataContext implementation which predates futil's method, but either way, it's about cascading the results of one search into the input of another so it seems like an intuitive name. Flow, pipe, or chain could work as well but those seem like they'd be worse names. It's more generic than just subquery but we could just call it that for now, although that complicates the eventual server side subquery implementation 🤔

Reply:
We could copy that explanation to give more context about why we picked the name. `cascade`, `pipe`, `chain` are generic words, but this implementation isn’t generic.